activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb

```diff
@@ -1156,13 +1172,12 @@ def extract_schema_qualified_name(string)
             [name.schema, name.identifier]
           end
 
-          def column_names_from_column_numbers(table_oid, column_numbers)
-            Hash[query(<<~SQL, "SCHEMA")].values_at(*column_numbers).compact
-              SELECT a.attnum, a.attname
-              FROM pg_attribute a
-              WHERE a.attrelid = #{table_oid}
-              AND a.attnum IN (#{column_numbers.join(", ")})
-            SQL
+          def decode_string_array(value)
+            array_decoder.decode(value)
+          end
+
+          def array_decoder
+            @array_decoder ||= PG::TextDecoder::Array.new
```

Comment:
Since there's a schema cache, we only really need this early, I don't think this memoization is worth it.

Might as well keep it as a local var in the method that need it like before even if it means one extra allocation.